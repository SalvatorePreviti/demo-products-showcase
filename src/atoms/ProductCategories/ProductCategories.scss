.item {
  display: inline-block;
  margin-left: 10px;
  padding: 5px;
  text-decoration: none;

  &,
  &:visited,
  &:active,
  &:focus {
    color: #000;
  }
}

.active {
  font-weight: bold;
  text-decoration: underline;
}
