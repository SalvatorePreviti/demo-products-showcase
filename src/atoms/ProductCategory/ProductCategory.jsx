import React from 'react'
import PropTypes from 'prop-types'
import styles from './ProductCategory.scss'

const itemStyle = styles.ProductCategory
const itemActiveStyle = `${itemStyle} ${styles.active}`

function ProductCategory(props) {
  const { title, active, onSelect } = props

  const onClick = event => {
    if (onSelect) {
      onSelect()
    }
    if (event) {
      event.preventDefault()
    }
  }

  return (
    <a href="#" className={active ? itemActiveStyle : itemStyle} onClick={onClick}>
      {title}
    </a>
  )
}

ProductCategory.propTypes = {
  title: PropTypes.string,
  active: PropTypes.bool,
  onSelect: PropTypes.func
}

export default ProductCategory
