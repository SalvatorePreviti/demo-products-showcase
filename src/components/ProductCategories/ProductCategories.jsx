import React from 'react'
import PropTypes from 'prop-types'
import styles from './ProductCategories.scss'

const itemStyle = styles.item
const itemActiveStyle = styles.item + ' ' + styles.active

function ProductCategories(props) {
  const { categories, activeCategoryId, onSelectCategory } = props
  return (
    <header>
      Store Cupboard
      {categories.map(category => {
        const categoryId = category.id
        return (
          <a
            key={categoryId}
            href="#"
            className={categoryId === activeCategoryId ? itemActiveStyle : itemStyle}
            onClick={() => onSelectCategory && onSelectCategory(categoryId)}>
            {category.title}
          </a>
        )
      })}
    </header>
  )
}

ProductCategories.propTypes = {
  categories: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.string.isRequired,
      title: PropTypes.string.isRequired
    })
  ).isRequired,
  activeCategoryId: PropTypes.string,
  onSelectCategory: PropTypes.func
}

export default ProductCategories
