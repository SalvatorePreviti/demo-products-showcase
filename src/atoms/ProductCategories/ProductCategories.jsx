import React from 'react'
import PropTypes from 'prop-types'
import ProductCategory from '../ProductCategory/ProductCategory'

function ProductCategories(props) {
  const { categories, activeCategory, onSelectCategory } = props

  return (
    <header>
      Store Cupboard
      {categories.map(category => (
        <ProductCategory
          key={category.id}
          title={category.title}
          active={category === activeCategory}
          onSelect={() => onSelectCategory && onSelectCategory(category)}
        />
      ))}
    </header>
  )
}

const productCategoryType = PropTypes.shape({
  id: PropTypes.string.isRequired,
  title: PropTypes.string.isRequired
})

ProductCategories.propTypes = {
  categories: PropTypes.arrayOf(productCategoryType.isRequired).isRequired,
  activeCategory: productCategoryType,
  onSelectCategory: PropTypes.func
}

export default ProductCategories
