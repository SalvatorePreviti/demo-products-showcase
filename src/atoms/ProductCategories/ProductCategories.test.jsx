import React from 'react'
import { shallow } from 'enzyme'

import ProductCategory from '../ProductCategory/ProductCategory'
import ProductCategories from './ProductCategories'

describe('ProductCategories', () => {
  const categories = [
    { id: '1', title: 'title 1' },
    { id: '2', title: 'title 2' },
    { id: '3', title: 'title 3' },
    { id: '4', title: 'title 4' }
  ]

  describe('with a list of categories, without selection', () => {
    let root

    beforeAll(() => {
      root = shallow(<ProductCategories categories={categories} />)
    })

    it('snapshots', () => {
      expect(root).toMatchSnapshot()
    })

    it('renders all elements', () => {
      const found = root.find(ProductCategory)
      expect(found.map(item => item.props())).toMatchObject([
        { active: false, title: categories[0].title },
        { active: false, title: categories[1].title },
        { active: false, title: categories[2].title },
        { active: false, title: categories[3].title }
      ])
    })
  })

  describe('with a list of categories, with a selection', () => {
    let root

    beforeAll(() => {
      root = shallow(<ProductCategories categories={categories} activeCategory={categories[1]} />)
    })

    it('snapshots', () => {
      expect(root).toMatchSnapshot()
    })

    it('renders all elements, marking as active the active element', () => {
      const found = root.find(ProductCategory)
      expect(found.map(item => item.props())).toMatchObject([
        { active: false, title: categories[0].title },
        { active: true, title: categories[1].title },
        { active: false, title: categories[2].title },
        { active: false, title: categories[3].title }
      ])
    })
  })

  describe('when clicking', () => {
    it('invokes onSelectCategory with the category id', () => {
      let clickedValue
      const root = shallow(
        <ProductCategories
          categories={categories}
          onSelectCategory={value => {
            clickedValue = value
          }}
        />
      )

      root
        .find(ProductCategory)
        .at(2)
        .simulate('select', categories[2])

      expect(clickedValue).toBe(categories[2])
    })
  })
})
