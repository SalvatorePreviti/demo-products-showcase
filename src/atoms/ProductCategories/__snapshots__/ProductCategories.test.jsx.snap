// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductCategories with a list of categories, with a selection snapshots 1`] = `
<header>
  Store Cupboard
  <ProductCategory
    active={false}
    key="1"
    onSelect={[Function]}
    title="title 1"
  />
  <ProductCategory
    active={true}
    key="2"
    onSelect={[Function]}
    title="title 2"
  />
  <ProductCategory
    active={false}
    key="3"
    onSelect={[Function]}
    title="title 3"
  />
  <ProductCategory
    active={false}
    key="4"
    onSelect={[Function]}
    title="title 4"
  />
</header>
`;

exports[`ProductCategories with a list of categories, without selection snapshots 1`] = `
<header>
  Store Cupboard
  <ProductCategory
    active={false}
    key="1"
    onSelect={[Function]}
    title="title 1"
  />
  <ProductCategory
    active={false}
    key="2"
    onSelect={[Function]}
    title="title 2"
  />
  <ProductCategory
    active={false}
    key="3"
    onSelect={[Function]}
    title="title 3"
  />
  <ProductCategory
    active={false}
    key="4"
    onSelect={[Function]}
    title="title 4"
  />
</header>
`;
